File: backend/app/schemas/job.py
"""Job request and response schemas."""

from __future__ import annotations

import uuid
from datetime import datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.models.enums import (
    EducationLevel,
    EmploymentType,
    ExperienceLevel,
    JobSource,
    JobStatus,
    SalaryPeriod,
    SkillRequirement,
    WorkMode,
)
from app.schemas.application import ApplicationRead
from app.schemas.urls import normalize_url

# Generous, because a real posting with a long benefits section runs long, and
# rejecting one for length is a worse failure than storing a few extra KB. The
# lower bound is enforced after cleaning, in the parser, where it can explain
# itself.
MAX_DESCRIPTION_CHARS = 60_000

# Wider than the profile's MAX_URL because application links are genuinely long:
# Greenhouse, Workday and Lever carry gh_jid, utm_* and referral tokens, and 500
# is reachable. The column is TEXT with no bound of its own, so this is purely a
# request-size guard.
MAX_JOB_URL = 1000


def _application_link(value: object) -> object:
    """Normalise and validate an application link.

    One function so the submit path and the add-a-link PATCH cannot drift; both
    end up in the same `href`. `required=True` because both fields are.
    """
    return normalize_url(value, max_length=MAX_JOB_URL, required=True)


def _optional_application_link(value: object) -> object:
    """The same rules, but a bad link costs the link rather than the record.

    Imports need both halves of this. Without any normalisation the importer is
    a route into an `href` that skips schemas/urls.py entirely — the exact hole
    the fetch path calls normalize_url explicitly to close — and it can store a
    value that is not a link, which the interface then shows as "no application
    link" while refusing to let anyone repair it.

    But raising here would reject the whole batch on one bad URL: this is a
    field validator on a list item, so the 422 kills every other record with it,
    and US-3.3 AC2 says a failed record must not abort the batch. Dropping the
    unusable link keeps the posting — whose skills and salary are what the
    ranking actually consumes — and the detail page then offers to add one.
    """
    try:
        return normalize_url(value, max_length=MAX_JOB_URL)
    except ValueError:
        return None


class CompanyRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    website: str | None = None
    industry: str | None = None


class JobSkillRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    skill_id: uuid.UUID
    name: str
    requirement: SkillRequirement
    min_years: Decimal | None = None
    extraction_confidence: Decimal | None = None


class JobSummary(BaseModel):
    """List-row shape. Deliberately excludes the description.

    A browse page of twenty jobs would otherwise ship twenty full postings —
    hundreds of kilobytes to render a list of titles.
    """

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    title: str
    company: CompanyRead | None = None
    location: str | None = None
    country_code: str | None = None
    work_mode: WorkMode | None = None
    employment_type: EmploymentType | None = None
    experience_level: ExperienceLevel | None = None
    min_years_experience: Decimal | None = None
    max_years_experience: Decimal | None = None
    salary_min: Decimal | None = None
    salary_max: Decimal | None = None
    salary_currency: str | None = None
    salary_period: SalaryPeriod | None = None
    posted_at: datetime | None = None
    created_at: datetime
    skill_count: int = 0
    #: What the *calling* user has done about this job, or null.
    #:
    #: A per-caller field on a shared-corpus row, which is worth justifying: both
    #: job endpoints already require a caller, there is no anonymous path and no
    #: cache header anywhere in this API, so the response is per-caller by
    #: construction and this only makes an existing property visible.
    #:
    #: A nullable object rather than `is_saved`/`is_applied` booleans. It reads
    #: as "a separate entity of yours, possibly absent" instead of as a property
    #: of the job; it carries applied_at, which the remove-confirmation needs and
    #: booleans have nowhere to put; and it is the same shape PUT returns, so the
    #: client swaps one field with no translation either way.
    application: ApplicationRead | None = None


class JobDetail(JobSummary):
    source: JobSource
    source_url: str | None = None
    status: JobStatus
    description_raw: str
    responsibilities: list[str] = Field(default_factory=list)
    requirements: list[str] = Field(default_factory=list)
    benefits: list[str] = Field(default_factory=list)
    min_education: EducationLevel | None = None
    expires_at: datetime | None = None
    skills: list[JobSkillRead] = Field(default_factory=list)


class SimilarJob(BaseModel):
    job: JobSummary
    #: Raw cosine in [0, 1]. Returned for evaluation and debugging and
    #: **deliberately not shown to the user**: ml.md is explicit that raw cosine
    #: on this model needs rescaling from its observed [0.3, 0.95] range before
    #: it means anything, and that rescaling belongs to the scoring step. A bare
    #: 0.71 on a card would look like a percentage and would not be one.
    similarity: float


class SimilarJobsResponse(BaseModel):
    """Nearest neighbours by embedding.

    `availability` exists because an empty `items` has three different meanings
    and the interface has to say which:

      READY    the comparison ran; nothing cleared the similarity floor
      PENDING  this posting has no vector yet, so there was nothing to compare
      DISABLED no embedding provider is configured at all (the default)
    """

    items: list[SimilarJob]
    availability: Literal["READY", "PENDING", "DISABLED"]
    limit: int
    #: Which vectors produced this, so a result is attributable — the same
    #: reason the rows carry it (ADR-007).
    model_name: str | None = None
    model_version: str | None = None


class JobListResponse(BaseModel):
    items: list[JobSummary]
    total: int
    limit: int
    offset: int


class JobSubmitRequest(BaseModel):
    """Paste a posting (US-3.1 AC1).

    `title` and `company` are optional overrides. Supplied, they win over
    anything the parser reads out of the text — the person pasting knows what
    they pasted, and a heuristic that overrules them is infuriating.
    """

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "description": "Senior Backend Engineer\n\nRequirements\n- 5+ years of Python...",
                "title": "Senior Backend Engineer",
                "company": "Acme Inc",
                "source_url": "https://example.com/jobs/123",
            }
        }
    )

    description: str = Field(min_length=1, max_length=MAX_DESCRIPTION_CHARS)
    title: str | None = Field(default=None, max_length=300)
    company: str | None = Field(default=None, max_length=200)
    # Required: this is the link the interface offers as "Apply for this job",
    # and a posting someone pasted always came from a page. Note that the
    # importer deliberately does NOT require one — see JobImportRecord.url.
    source_url: str = Field(max_length=MAX_JOB_URL)

    # source_url is deliberately NOT in this list. Blanking it to None would
    # reach Pydantic as a missing str and surface as "Input should be a valid
    # string"; _apply_link says what is actually wrong instead.
    @field_validator("title", "company", mode="before")
    @classmethod
    def _blank_to_none(cls, value: object) -> object:
        # A cleared input posts "", which would otherwise override the parser
        # with an empty string and produce a job titled "".
        if isinstance(value, str) and not value.strip():
            return None
        return value

    _apply_link = field_validator("source_url", mode="before")(_application_link)


class JobSubmitResponse(BaseModel):
    job: JobDetail
    is_duplicate: bool = Field(
        description=(
            "True when this posting was already in the corpus; the existing job is returned."
        )
    )


class JobApplicationLinkUpdate(BaseModel):
    """Attach an application link to a job that has none."""

    source_url: str = Field(max_length=MAX_JOB_URL)

    _apply_link = field_validator("source_url", mode="before")(_application_link)


class JobImportRecord(BaseModel):
    external_id: str | None = Field(default=None, max_length=200)
    title: str | None = Field(default=None, max_length=300)
    company: str | None = Field(default=None, max_length=200)
    description: str = Field(min_length=1, max_length=MAX_DESCRIPTION_CHARS)
    # Optional, unlike JobSubmitRequest.source_url, and that divergence is
    # deliberate rather than an oversight. A dataset row's value to the ranking
    # formula — skills, salary, requirements — does not depend on a link, and
    # failing a whole record over a missing one makes an operator bisect the
    # file for a non-reason. A linkless imported row is exactly what the
    # "add the application link" control on the detail page exists for.
    url: str | None = Field(default=None, max_length=MAX_JOB_URL)
    # Structured metadata beats parsing it back out of prose. country_code has
    # no parser at all, so without this the documented `GET /jobs?country_code=`
    # filter matches nothing.
    location: str | None = Field(default=None, max_length=200)
    country_code: str | None = Field(default=None, max_length=2)
    posted_at: datetime | None = None

    _link = field_validator("url", mode="before")(_optional_application_link)


class JobImportRequest(BaseModel):
    # Bounded so one request cannot hold a transaction open across a whole
    # dataset. A larger file is imported as several batches.
    records: list[JobImportRecord] = Field(min_length=1, max_length=500)


class ImportFailureRead(BaseModel):
    index: int
    external_id: str | None = None
    reason: str


class JobImportResponse(BaseModel):
    """A partial success is the normal outcome, not an error (US-3.3 AC2)."""

    created: int
    duplicates: int
    failed: list[ImportFailureRead]
    processed: int


class JobFetchRequest(BaseModel):
    query: str = Field(min_length=2, max_length=200)
    country: str = Field(default="in", min_length=2, max_length=2)
    # Defaults to one, and that is the whole answer to an accidental
    # double-fire: on a free tier measured in a few hundred calls a month, a
    # double-click must cost one request rather than five. Five is the ceiling
    # and choosing it is a deliberate act. api.md section 1.8 specifies an
    # Idempotency-Key for this class of endpoint; nothing implements one yet
    # (Redis is configured but unused), so this plus (source, external_id)
    # dedup is the interim — a repeat spends quota but creates no rows.
    max_pages: int = Field(default=1, ge=1, le=5)
    #: Only postings this recent. Without it a repeated fetch returns the same
    #: relevance-ranked results and spends quota on rows already in the corpus.
    posted_within_days: int | None = Field(default=None, ge=1, le=90)


class JobFetchResponse(BaseModel):
    """Deliberately the same report shape as JobImportResponse.

    Standalone rather than a subclass because that model's docstring is US-3.3's
    contract and should stay that, but `failed` is the identical type: an
    operator reading either one is asking the same question.
    """

    provider: str
    created: int
    duplicates: int
    failed: list[ImportFailureRead]
    processed: int
    pages_fetched: int
    postings_seen: int
    stopped_early: bool
    stop_reason: str | None = None
    #: Surfaced in the response, not just the log: "what did that cost me" is a
    #: question asked at the moment the button is pressed.
    quota_remaining: int | None = None

File: frontend/src/types/job.ts
/** Job types mirroring backend/app/schemas/job.py. */

import type { ApplicationRead } from '@/types/application'
import type { EmploymentType, WorkMode } from '@/types/profile'

export type ExperienceLevel =
  | 'INTERN'
  | 'ENTRY'
  | 'JUNIOR'
  | 'MID'
  | 'SENIOR'
  | 'LEAD'
  | 'PRINCIPAL'

export type EducationLevel =
  | 'NONE'
  | 'HIGH_SCHOOL'
  | 'DIPLOMA'
  | 'BACHELORS'
  | 'MASTERS'
  | 'DOCTORATE'

export type SalaryPeriod = 'YEARLY' | 'MONTHLY' | 'HOURLY'
export type SkillRequirement = 'REQUIRED' | 'PREFERRED'
export type JobSource = 'USER_SUBMITTED' | 'DATASET_IMPORT' | 'PARTNER_API'
export type JobStatus = 'ACTIVE' | 'DUPLICATE'

export interface CompanyRead {
  id: string
  name: string
  website: string | null
  industry: string | null
}

export interface JobSkillRead {
  skill_id: string
  name: string
  requirement: SkillRequirement
  /** Numerics arrive as strings so no precision is lost in transit. */
  min_years: string | null
  extraction_confidence: string | null
}

export interface JobSummary {
  id: string
  title: string
  company: CompanyRead | null
  location: string | null
  country_code: string | null
  work_mode: WorkMode | null
  employment_type: EmploymentType | null
  experience_level: ExperienceLevel | null
  min_years_experience: string | null
  max_years_experience: string | null
  salary_min: string | null
  salary_max: string | null
  salary_currency: string | null
  salary_period: SalaryPeriod | null
  posted_at: string | null
  created_at: string
  skill_count: number
  /**
   * What *you* have done about this job, or null.
   *
   * Per-caller, on a row from a shared corpus — which is fine here because both
   * job endpoints already require a caller and nothing is cached. A nullable
   * object rather than `is_saved`/`is_applied` booleans: it reads as a separate
   * entity of yours rather than a property of the job, it carries `applied_at`
   * which the remove-confirmation needs, and it is the same shape the PUT
   * returns, so updating a row is a field swap with no translation.
   */
  application: ApplicationRead | null
  /**
   * Your match score out of 100, when the list was sorted by match.
   *
   * Null everywhere else — and null is not zero. A browse-sorted list has not
   * computed a score, which is a different statement from "this job scores
   * nothing for you". Render the row without a score rather than a 0.
   *
   * A string, like every score in this API, so 68.4 cannot arrive as
   * 68.40000000000001.
   */
  match_score: string | null
}

export interface JobDetail extends JobSummary {
  source: JobSource
  source_url: string | null
  status: JobStatus
  description_raw: string
  responsibilities: string[]
  requirements: string[]
  benefits: string[]
  min_education: EducationLevel | null
  expires_at: string | null
  skills: JobSkillRead[]
}

export interface JobListResponse {
  items: JobSummary[]
  /**
   * Rows in the whole list.
   *
   * Under `sort=match` this is **capped by the recall limit (200)**, so it is the
   * number of ranked matches rather than a corpus count. The page's wording has
   * to reflect that: "247 matches" would be an invented number.
   */
  total: number
  limit: number
  offset: number
  /**
   * Why `items` may be empty under `sort=match`.
   *
   * Always READY for a date-sorted browse, which cannot fail these ways. The two
   * other values are not errors: NO_RESUME means there is nothing to match
   * against, PENDING means the resume is not embedded yet. Both arrive as 200s
   * with data, because an empty list with no explanation reads as "no job in the
   * world suits you" — a much more discouraging claim than either true one.
   */
  availability: 'READY' | 'PENDING' | 'NO_RESUME'
  /** Which ranking produced the scores, or null in browse mode. */
  ranking_version: string | null
}

export interface JobSubmitResponse {
  job: JobDetail
  is_duplicate: boolean
}

/**
 * Link state that JobDetailPage understands.
 *
 * Two independent senders — JobCard, which knows which list you were looking
 * at, and AddJobPage, which knows the posting was already in the corpus. Both
 * fields are optional and read independently, so a third sender adds a field
 * rather than colliding with the other two.
 */
export interface JobDetailLocationState {
  /** Path and search of the list to return to, e.g. "/jobs?q=python&offset=20". */
  backTo?: string
  isDuplicate?: boolean
}

export interface SimilarJob {
  job: JobSummary
  /**
   * Raw cosine in [0, 1], for debugging and evaluation — **never rendered**.
   * Raw cosine on this model needs rescaling before it means anything (ml.md
   * §4.1), and that rescaling belongs to the scoring step. A bare 0.71 on a
   * card would look like a percentage and would not be one.
   */
  similarity: number
}

export interface SimilarJobsResponse {
  items: SimilarJob[]
  /**
   * Why `items` is empty, because it has three different reasons:
   *   READY    the comparison ran and nothing was close enough
   *   PENDING  this posting has no vector yet
   *   DISABLED embeddings are switched off entirely (the default)
   */
  availability: 'READY' | 'PENDING' | 'DISABLED'
  limit: number
  model_name: string | null
  model_version: string | null
}

export interface JobFilters {
  q?: string
  work_mode?: WorkMode
  employment_type?: EmploymentType
  /**
   * Still a supported API filter, though the browse UI now filters on years —
   * the ranking formula is numeric and the seniority enum plays no part in it.
   */
  experience_level?: ExperienceLevel
  /** Show jobs whose stated range covers this many years. */
  years_experience?: number
  country_code?: string
  /**
   * Show postings published within this many days. Undated postings are
   * included — over half the corpus has no stated date, and a missing date says
   * nothing about age.
   */
  posted_within_days?: number
  limit?: number
  offset?: number
}

/** Mirrors MAX_DESCRIPTION_CHARS in backend/app/schemas/job.py. */
export const MAX_DESCRIPTION_CHARS = 60_000
/** Mirrors MIN_DESCRIPTION_CHARS in backend/app/services/job/pipeline.py. */
export const MIN_DESCRIPTION_CHARS = 200

/**
 * Preset choices for the browse filter. `value` is the number sent as
 * `years_experience`; a job matches when its stated range covers it.
 *
 * Generated rather than written out: eleven near-identical entries invite a
 * typo, and the only irregularity is the singular at 1.
 *
 * No `keywords` needed — comboboxCore's normalizeText folds "5+ years" to
 * "5 years", so typing 5 is already a prefix match, and typing 1 correctly
 * offers both "1+ year" and "10+ years".
 */
export const EXPERIENCE_YEAR_OPTIONS = Array.from({ length: 11 }, (_, years) => ({
  value: String(years),
  label: years === 1 ? '1+ year' : `${years}+ years`,
}))

/**
 * Preset windows for the "Posted within" filter. `value` is the day count sent
 * as `posted_within_days`.
 *
 * Deliberately coarse. The provider reports whole days and often no date at
 * all, so offering "last 24 hours" would imply a resolution the data has not
 * got. Thirty days is the widest because a posting older than that is rarely
 * still open, and the default — no window — already shows everything.
 */
export const POSTED_WITHIN_OPTIONS = [
  { value: '7', label: 'Last 7 days' },
  { value: '14', label: 'Last 14 days' },
  { value: '30', label: 'Last 30 days' },
]

const EDUCATION_LABELS: Record<EducationLevel, string> = {
  NONE: 'No formal requirement',
  HIGH_SCHOOL: 'High school',
  DIPLOMA: 'Diploma',
  BACHELORS: "Bachelor's degree",
  MASTERS: "Master's degree",
  DOCTORATE: 'Doctorate',
}

export function educationLabel(level: EducationLevel): string {
  return EDUCATION_LABELS[level]
}

/** "Full time" from "FULL_TIME" — for enums with no curated label list. */
export function humanise(value: string): string {
  return value.charAt(0) + value.slice(1).toLowerCase().replace(/_/g, ' ')
}

const PERIOD_SUFFIX: Record<SalaryPeriod, string> = {
  YEARLY: '/yr',
  MONTHLY: '/mo',
  HOURLY: '/hr',
}

/**
 * Render a pay range the way the posting stated it.
 *
 * Deliberately does not convert currencies or normalise periods. The backend
 * stores what the posting said, and inventing a converted figure would put a
 * number in front of the user that no employer ever wrote.
 */
export function formatSalary(job: {
  salary_min: string | null
  salary_max: string | null
  salary_currency: string | null
  salary_period: SalaryPeriod | null
}): string | null {
  if (job.salary_min === null && job.salary_max === null) return null

  const currency = job.salary_currency ?? ''
  const format = (value: string) => {
    const amount = Number(value)
    if (!Number.isFinite(amount)) return value
    // Compact notation: a range of 2,800,000–4,500,000 is unreadable at a
    // glance, and "2.8M" is what the eye is actually looking for in a list.
    return new Intl.NumberFormat(undefined, {
      notation: amount >= 100_000 ? 'compact' : 'standard',
      maximumFractionDigits: 1,
    }).format(amount)
  }

  const range =
    job.salary_min !== null && job.salary_max !== null
      ? `${format(job.salary_min)} – ${format(job.salary_max)}`
      : format((job.salary_min ?? job.salary_max)!)

  const suffix = job.salary_period !== null ? PERIOD_SUFFIX[job.salary_period] : ''
  return `${currency} ${range}${suffix}`.trim()
}

/** "4–7 years", "5+ years", or null when the posting did not say. */
export function formatExperience(job: {
  min_years_experience: string | null
  max_years_experience: string | null
}): string | null {
  const min = job.min_years_experience
  const max = job.max_years_experience
  if (min === null && max === null) return null

  const n = (value: string) => Number(value).toString()
  if (min !== null && max !== null) return `${n(min)}–${n(max)} years`
  if (min !== null) return `${n(min)}+ years`
  return `up to ${n(max!)} years`
}

// ---------------------------------------------------------------- match score

/** The six dimensions of the ranking formula, in documented weight order. */
export type MatchDimensionName =
  | 'semantic'
  | 'skill'
  | 'experience'
  | 'education'
  | 'location'
  | 'salary'

/**
 * Why a dimension scored what it did.
 *
 * The distinction that matters to the interface is the last two. Both score a
 * neutral 0.5, but `NEEDS_PROFILE` is something the user can fix and
 * `NEEDS_DATA` is not — showing "complete your profile" because an employer
 * left a field blank would be blaming the reader for someone else's omission.
 */
export type MatchDimensionStatus = 'SCORED' | 'NOT_STATED' | 'NEEDS_PROFILE' | 'NEEDS_DATA'

export interface MatchDimension {
  dimension: MatchDimensionName
  /** [0, 1]. Decimal over the wire, so a string. */
  score: string
  weight: string
  /** `score × weight × 100`, to one place. The six sum to `overall_score`. */
  contribution: string
  status: MatchDimensionStatus
  reason: string
}

export interface MatchedSkill {
  id: string
  name: string
  requirement: 'REQUIRED' | 'PREFERRED'
}

export interface MatchResponse {
  job_id: string
  /**
   *   READY      all six dimensions ran
   *   PARTIAL    the semantic dimension could not run — no provider, or a
   *              vector that is not built yet. The breakdown still sums
   *   NO_RESUME  nothing uploaded, so there is no score at all
   */
  availability: 'READY' | 'PARTIAL' | 'NO_RESUME'
  overall_score: string | null
  breakdown: MatchDimension[]
  /** Weight of the rows that measured something — the honesty disclosure. */
  scored_weight: string
  skills: {
    matched: MatchedSkill[]
    partial: MatchedSkill[]
    missing: MatchedSkill[]
  }
  ranking_version: string | null
  resume_version_id: string | null
  computed_at: string | null
}

// ------------------------------------------------------------ recommendations

export interface RecommendedJob {
  job: JobSummary
  /** 0–100, one decimal place. Decimal over the wire, so a string. */
  score: string
  breakdown: MatchDimension[]
  scored_weight: string
  skills: {
    matched: MatchedSkill[]
    partial: MatchedSkill[]
    missing: MatchedSkill[]
  }
}

export interface RecommendationsResponse {
  items: RecommendedJob[]
  /**
   *   READY      the ranking ran
   *   PENDING    the resume has no vector yet — the indexer has not reached it
   *   NO_RESUME  nothing uploaded, so there is nothing to rank against
   */
  availability: 'READY' | 'PENDING' | 'NO_RESUME'
  /** Opaque. Null on the last page — never parse it. */
  next_cursor: string | null
  limit: number
  /**
   * How many of the recalled set survived the filters. Deliberately not a
   * corpus-wide total: two-stage retrieval never looks at the whole corpus, so
   * any "total" would be a number about the recall set dressed as a number
   * about the market.
   */
  considered: number
  ranking_version: string | null
  resume_version_id: string | null
  computed_at: string | null
}
